### PR TITLE
Use faster UTF8 encoding in `Content.write()`

### LINKED ISSUE #12469: Content.Sink.write(sink, last, utf8Content, callback) could become faster
**Jetty version(s)**
Jetty 12.0.x

**Enhancement Description**
Looking at the source code, I see that the [method body](https://github.com/jetty/jetty.project/blob/1e4eb6151eb79d0ebdc1a4df167a323a2c3dc344/jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java#L695) of `Content.Sink.write(sink, last, utf8Content, callback)` is using `StandardCharsets.UTF_8.encode(utf8Content)` to create the UTF-8 ByteBuffer.

I believe that `ByteBuffer.wrap(utf8Content.getBytes(StandardCharsets.UTF_8))` would be quite faster, as in recent Java versions (which use UTF8 internally) it is implemented as a simple `System.arrayCopy()`.

### PR DESCRIPTION
**Pull request by lorban**
Fixes #12469

### CODE REVIEW

File: tests/jetty-jmh/src/main/java/org/eclipse/jetty/io/jmh/Utf8Benchmark.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+@State(Scope.Benchmark)
+@Threads(4)
+@Warmup(iterations = 7, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 7, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+public class Utf8Benchmark
+{
+    private static final String ASCII_STRING = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor " +
```
Comment (sbordet) [inline tests/jetty-jmh/src/main/java/org/eclipse/jetty/io/jmh/Utf8Benchmark.java]: Use a text block, with EOL backslash if you don't want newlines.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java [hunk 2]
```diff
@@ -692,7 +692,7 @@ static void write(Sink sink, boolean last, ByteBuffer byteBuffer) throws IOExcep
          */
         static void write(Sink sink, boolean last, String utf8Content, Callback callback)
         {
-            sink.write(last, StandardCharsets.UTF_8.encode(utf8Content), callback);
+            sink.write(last, ByteBuffer.wrap(utf8Content.getBytes(StandardCharsets.UTF_8)), callback);
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:695]: Feels like a lot of data copying and allocation, and something that may even depend on the Java version.

What Java versions have you tried? It is the case that the performance is that bad in all recent Java versions?

**Review by sbordet — CHANGES_REQUESTED** [review]

**Review by joakime — CHANGES_REQUESTED** [review]
Please confirm that the behavior when dealing with bad input is the same with the new code vs the old code.

Example:
A bad / Malformed input String that would trigger a fault in the conversion.
If the old code throws an exception, this new one should too.
If the old code used replacement characters, this new one should too.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:695]: I've tried both JDK 17 and 22 and they have similar performance, which can be summed up as following:

 - when the string is pure ASCII, wrap/getBytes just a memcopy and is ridiculously faster than encode, about 20X faster! Profiling the allocation rate shows that mem bandwidth is the bottleneck.
 - the more you move away from pure ASCII, the lesser the perf benefits become: about 2.5X faster for FR, about 50% faster for JA. Allocations also increase: 5X for FR, 20% for JA.

**Comment by lorban** [comment]
@joakime is it possible to create a `String` object containing invalid UTF8? All I've seen in our tests is to create invalid UTF8 in byte arrays then the `String` constructor is used to build the string, which does apply some correction. Did I get that right?

**Review by sbordet — DISMISSED** [review]

**Review by joakime — APPROVED** [review]
It is difficult to create a String that CharsetDecoder.encode(String) would fail on.

**Comment by lorban** [comment]
I found one other place where we use `encode` that could be replaced with `getBytes/wrap`.

**Comment by lorban** [comment]
For future reference, here is the benchmark's report:

```
Benchmark                                          (locale)   Mode  Cnt         Score         Error   Units
Utf8Benchmark.testEncode                              ASCII  thrpt   10   1885900.209 ±   12517.384   ops/s
Utf8Benchmark.testEncode:gc.alloc.rate                ASCII  thrpt   10      1121.696 ±       7.530  MB/sec
Utf8Benchmark.testEncode:gc.alloc.rate.norm           ASCII  thrpt   10       624.007 ±       0.001    B/op
Utf8Benchmark.testEncode:gc.count                     ASCII  thrpt   10        19.000                counts
Utf8Benchmark.testEncode:gc.time                      ASCII  thrpt   10        23.000                    ms
Utf8Benchmark.testEncode                                 FR  thrpt   10   1310399.805 ±   12798.866   ops/s
Utf8Benchmark.testEncode:gc.alloc.rate                   FR  thrpt   10       789.489 ±       7.739  MB/sec
Utf8Benchmark.testEncode:gc.alloc.rate.norm              FR  thrpt   10       632.011 ±       0.001    B/op
Utf8Benchmark.testEncode:gc.count                        FR  thrpt   10        14.000                counts
Utf8Benchmark.testEncode:gc.time                         FR  thrpt   10        18.000                    ms
Utf8Benchmark.testEncode                                 JA  thrpt   10    814449.918 ±   11152.653   ops/s
Utf8Benchmark.testEncode:gc.alloc.rate                   JA  thrpt   10      2925.414 ±      40.086  MB/sec
Utf8Benchmark.testEncode:gc.alloc.rate.norm              JA  thrpt   10      3768.017 ±       0.001    B/op
Utf8Benchmark.testEncode:gc.count                        JA  thrpt   10        33.000                counts
Utf8Benchmark.testEncode:gc.time                         JA  thrpt   10        47.000                    ms
Utf8Benchmark.testWrapGetBytes                        ASCII  thrpt   10  39417563.752 ± 1256275.047   ops/s
Utf8Benchmark.testWrapGetBytes:gc.alloc.rate          ASCII  thrpt   10     19538.322 ±     623.689  MB/sec
Utf8Benchmark.testWrapGetBytes:gc.alloc.rate.norm     ASCII  thrpt   10       520.000 ±       0.001    B/op
Utf8Benchmark.testWrapGetBytes:gc.count               ASCII  thrpt   10        71.000                counts
Utf8Benchmark.testWrapGetBytes:gc.time                ASCII  thrpt   10       144.000                    ms
Utf8Benchmark.testWrapGetBytes                           FR  thrpt   10   3434889.274 ±   64716.469   ops/s
Utf8Benchmark.testWrapGetBytes:gc.alloc.rate             FR  thrpt   10      4819.736 ±      90.934  MB/sec
Utf8Benchmark.testWrapGetBytes:gc.alloc.rate.norm        FR  thrpt   10      1472.004 ±       0.001    B/op
Utf8Benchmark.testWrapGetBytes:gc.count                  FR  thrpt   10        37.000                counts
Utf8Benchmark.testWrapGetBytes:gc.time                   FR  thrpt   10        58.000                    ms
Utf8Benchmark.testWrapGetBytes                           JA  thrpt   10   1399081.733 ±   47082.158   ops/s
Utf8Benchmark.testWrapGetBytes:gc.alloc.rate             JA  thrpt   10      3595.402 ±     121.188  MB/sec
Utf8Benchmark.testWrapGetBytes:gc.alloc.rate.norm        JA  thrpt   10      2696.010 ±       0.001    B/op
Utf8Benchmark.testWrapGetBytes:gc.count                  JA  thrpt   10        32.000                counts
Utf8Benchmark.testWrapGetBytes:gc.time                   JA  thrpt   10        46.000                    ms
```

**Review by gregw — APPROVED** [review]
Is there anything that can be done to improve org.eclipse.jetty.ee11.servlet.HttpOutput#print(java.lang.String, boolean)?

**Comment by lorban** [comment]
@gregw `HttpOutput.print()` goes into great length to pool the encoder and to to detect encoding errors like overflows/underflows.

We could theoretically replace all that with a much simpler `String.getBytes(Charset)`, which could improve perf but may not work as expected w.r.t encoding. @joakime what's your opinion on that one?

File: tests/jetty-jmh/src/main/java/org/eclipse/jetty/io/jmh/Utf8Benchmark.java (hunk 1 above)
Comment (lorban) [reply tests/jetty-jmh/src/main/java/org/eclipse/jetty/io/jmh/Utf8Benchmark.java]: Done.

**Comment by joakime** [comment]
> @joakime what's your opinion on that one?

If the behavior of the API to the users is maintained, then I'm in favor of the change.
Is it possible to use HttpOutput.print() with partial code points? (meaning a print() is called which starts the code points, then a subsequent print() results in finishing the code point)
If so, then the `String.getBytes(Charset)` wouldn't work for us.

**Comment by lorban** [comment]
I'm going to give `HttpOutput.print()` a try in another PR, as it isn't trivial to change but may work and be worth the effort.

**Pull request merged**